deploy

```diff
@@ -0,0 +1,192 @@
+
+log-header() {
+  declare desc="Log header formatter";
+  echo "====> $*"
+}
+
+log-info() {
+  declare desc="Log info formatter";
+  echo "      $*"
+}
+
+log-warn() {
+  declare desc="Log warn formatter";
+  echo " ?    $*" 1>&2
+}
+
+log-fail() {
+  declare desc="Log fail formatter";
+  echo " !    $*" 1>&2
+  exit 1
+}
+
+log-exit() {
+  declare desc="Log exit formatter";
+  echo " !    $*" 1>&2
+  exit 0
+}
+
+fn-check-env() {
+  declare APP="$1"
+
+  if grep -Fxq "$APP" "$SOURCE_DIR/.deployignore"; then
+    log-exit "App $APP is in the .deployignore file, skipping deploy"
+  fi
+
+  if ! command -v "dds-client" >/dev/null; then
+    log-fail "dds-client binary not found"
+  fi
+
+  if [[ -z "$DASH_ENTERPRISE_API_KEY" ]]; then
+    log-fail "DASH_ENTERPRISE_API_KEY is not defined"
+  fi
+
+  if [[ -z "$DASH_ENTERPRISE_USERNAME" ]]; then
+    log-fail "DASH_ENTERPRISE_USERNAME is not defined"
+  fi
+
+  if [[ -z "$DASH_ENTERPRISE_URL" ]]; then
+    log-fail "DASH_ENTERPRISE_URL is not defined"
+  fi
+
+  if [[ -z "$CIRCLE_SHA1" ]]; then
+    log-fail "CIRCLE_SHA1 is not defined"
+  fi
+}
+
+main() {
+  declare APP="$1"
+  local app_dir="$APPS_DIR/$APP"
+  local remote_url="$DASH_ENTERPRISE_URL/GIT/$APP"
+  local is_dashr=false
+  local tmp_shallow_clone="$(mktemp -d -t apps-XXXXXXXXXX)"
+  local tmp_full_clone="$(mktemp -d -t apps-XXXXXXXXXX)"
+  local force_push=false
+  local push_code=false
+  local exit_code remote_sha
+  trap "rm -rf '$tmp_shallow_clone' '$tmp_full_clone' >/dev/null" RETURN INT TERM EXIT
+
+  fn-check-env "$APP"
+
+  if [[ -z "$APP" ]]; then
+    log-fail "No app name specified"
+  fi
+
+  if [[ ! -d "$app_dir" ]]; then
+    log-fail "Invalid app name: $APP"
+  fi
+
+  if [[ "$APP" =~ "dashr" ]]; then
+    is_dashr=true
+    if [[ "$ALLOW_DASHR" != "true" ]]; then
+      log-exit "App $APP is a dashr app, skipping deploy"
+    fi
+  fi
+
+  if [[ ! -d "$app_dir/assets" ]]; then
+    log-fail "App $APP is missing the assets directory"
+  fi
+
+  log-header "Deploying $APP"
+  log-info "dashr: ${is_dashr}"
+  if [[ "$is_dashr" == "false" ]]; then
+    local celery_version="$(cat "$app_dir/requirements.txt" | grep -E "^celery=")"
+    local redis_version="$(cat "$app_dir/requirements.txt" | grep -E "^redis=")"
+
+    log-info "dash python version: $(cat "$app_dir/requirements.txt" | grep -E "^dash=")"
+    [[ -n "$celery_version" ]] && log-info "celery python version: $(cat "$app_dir/requirements.txt" | grep -E "^celery=")"
+    [[ -n "$redis_version" ]] && log-info "redis python version: $(cat "$app_dir/requirements.txt" | grep -E "^redis=")"
+
+    if [[ -n "$celery_version" ]] || [[ -n "$redis_version" ]]; then
+      log-exit "Skipping service that requires a redis instance"
+    fi
+  fi
+
+  if ! dds-client apps:exists --name "$APP" >/dev/null 2>&1; then
+    log-info "exists: false"
+    if [[ "$CREATE_APP" == "true" ]]; then
+      log-warn "$APP not found, creating"
+      dds-client apps:create --name "$APP"
```

Comment:
Nice! So much faster than going to the web interface to create an app :D

Reply:
Yeah and hopefully at some point I add support for creating redis/postgres for the apps too!